### PR TITLE
PARQUET-2374: Add metrics support for parquet file reader

### PR DESCRIPTION
**Pull request by parthchandra**
Make sure you have checked _all_ steps below.

### Jira

This PR addresses the following: [PARQUET-2374](https://issues.apache.org/jira/browse/PARQUET-2374)
Add metrics support for parquet file reader
### Tests
No new tests. Adds a new public interface for reporting metrics from the parquet file reader

### CODE REVIEW

**Comment by parthchandra** [comment]
Note that this can be used to compare read performance of the existing file reader with the the hadoop vector io reader https://github.com/apache/parquet-mr/pull/1139

**Comment by parthchandra** [comment]
@shangxinli could you take a look at this PR?

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ColumnChunkPageReadStore.java
Comment (wgtmac) [inline parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ColumnChunkPageReadStore.java:249]: ditto

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ColumnChunkPageReadStore.java
Comment (wgtmac) [inline parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ColumnChunkPageReadStore.java:259]: Do we need to include decompression of dictionary page?

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileReader.java
Comment (wgtmac) [inline parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileReader.java:1899]:
```suggestion
        LOG.debug(
          "Parquet: File Read stats:  Length: {} MB, Time: {} secs, throughput: {} MB/sec ",
          sizeInMb, timeInSec, throughput);
```

**Review by wgtmac — CHANGES_REQUESTED** [review]
Thanks for adding this! I have left a few comments.

**Review by parthchandra — COMMENTED** [review]
Thank you for reviewing @wgtmac !
Addressed your comments. There was some weird reformatting that occurred. I've fixed those in places you pointed out as well as some others.

**Review by wgtmac — APPROVED** [review]
Thanks @parthchandra! LGTM

**Comment by parthchandra** [comment]
Thank you @wgtmac !

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileReader.java
Comment (ConeyLiu) [inline parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileReader.java:138]: why the `synchronized` is needed?

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileReader.java
Comment (ConeyLiu) [inline parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileReader.java:128]: do we need this field as `public`?

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ColumnChunkPageReadStore.java
Comment (ConeyLiu) [inline parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ColumnChunkPageReadStore.java:83]: It should be `private final`

**Comment by steveloughran** [comment]
it'd be really nice if somehow there was a way to push hadoop stream IOStats here, especially the counters, min, max and mean maps: https://hadoop.apache.org/docs/current/hadoop-project-dist/hadoop-common/filesystem/iostatistics.html

and its really interesting for s3, azure and gcs clients, where we collect stream specific stuff, including things like: bytes discarded in seek, time for GET, whether we did a HEAD first, and more. These are collected in a thread level, but also include stats from helper threads such as those in async stream draining, vector IO...

It'd take a move to hadoop 3.3.1+ to embrace the API, but if there was a way for something to publish stats to your metric collector, then maybe something could be done

Tip: you can enable a dump of a filesystem's aggregate stats in process shutdown for azure and s3a
```
fs.iostatistics.logging.level=info
```

```
2023-11-17 18:30:28,634 [shutdown-hook-0] INFO  statistics.IOStatisticsLogging (IOStatisticsLogging.java:logIOStatisticsAtLevel(269)) - IOStatistics: counters=((action_http_head_request=3)
(audit_request_execution=15)
(audit_span_creation=12)
(object_list_request=12)
(object_metadata_request=3)
(op_get_file_status=1)
(op_glob_status=1)
(op_list_status=9)
(store_io_request=15));

gauges=();

minimums=((action_http_head_request.min=22)
(object_list_request.min=25)
(op_get_file_status.min=1)
(op_glob_status.min=9)
(op_list_status.min=25));

maximums=((action_http_head_request.max=41)
(object_list_request.max=398)
(op_get_file_status.max=1)
(op_glob_status.max=9)
(op_list_status.max=408));

means=((action_http_head_request.mean=(samples=3, sum=87, mean=29.0000))
(object_list_request.mean=(samples=12, sum=708, mean=59.0000))
(op_get_file_status.mean=(samples=1, sum=1, mean=1.0000))
(op_glob_status.mean=(samples=1, sum=9, mean=9.0000))
(op_list_status.mean=(samples=9, sum=814, mean=90.4444)));
```

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileReader.java
Comment (steveloughran) [inline parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileReader.java:1854]: s3a and abfs do lazy seek; no IO until the first read/readFully. best to measure that.

**Review by steveloughran — COMMENTED** [review]
For the object stores, things to measure are
* time to open() and close() a file
* time for a read after a backwards seek
* time for a read after a forwards seek.
* how many reads actually took place
* for vector IO, whatever gets picked up there
* were errors reported and retried, or throttling events
* number of underlying GET requests

as well as publishing this through the IOStatisticsSource API, we also now collect it at the thread level (IOStatisticsContext) with the goal of having the execution engine collect all stats for a task and then aggregate across the job. you can set the s3a committers up to do this.

If you were for a recent hadoop release only I'd say "publish the stream stats to the thread context for aggregation", but that's not viable here. What would be good if this stats was set up to
* take maps of key-value rather than a fixed enum
* collect those min/mean/max as well as counts.
* somehow provided a plugin point where we could add something to add any of the parquet reader/writer stats to the thread context -trying to collect stats from inside wrapped-many-times-over streams and iterators is way too complex. I know, i have a branch of parquet where I tried that...
*

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ColumnChunkPageReadStore.java
Comment (parthchandra) [reply parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ColumnChunkPageReadStore.java:83]: Changed

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileReader.java
Comment (parthchandra) [reply parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileReader.java:128]: No we don't. Made it private

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileReader.java
Comment (parthchandra) [reply parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileReader.java:138]: My mistake. I initially implemented the metrics callback as a singleton. The method no longer needs to be synchronized.

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileReader.java
Comment (parthchandra) [reply parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileReader.java:1854]: Right. Generally, seeks (especially backwards) cause the file system to stop their read-ahead and turn off sequential read optimizations. The seek call itself doesn't take much time.

**Comment by parthchandra** [comment]
@steveloughran I did look into leveraging Hadoop io stats  but my first attempt did not work too well and I thought a simpler initial implementation would be more useful. Once we move to hadoop vector io, I'll take another stab at it. 

> What would be good if this stats was set up to
> 
> take maps of key-value rather than a fixed enum

The fixed enum here is simply the Parquet file reader providing information that these are the values it knows about. This implementation is not really collecting and aggregating anything, it is simply recording the time and counts and passing them on. 
 
> collect those min/mean/max as well as counts.

The implementation of the parquet metrics callback will do that. So if the execution engine is Spark, it can simply get the values and add them to it's own metrics collection subsystem which then computes the min/max/mean.

> somehow provided a plugin point where we could add something to add any of the parquet reader/writer stats to the thread context -trying to collect stats from inside wrapped-many-times-over streams and iterators is way too complex. I know, i have a branch of parquet where I tried that...

Hmm, that will take some work. I wanted to measure streaming decompression time (where the `decompress` call simply returns a stream which is decompressed as it is read), but found it required too many breaking changes to implement. But a standard system like `IOStatistics` where such a stream is a IOStatisticsSource would be perfect.

**Comment by wgtmac** [comment]
> For the object stores, things to measure are
> 
> * time to open() and close() a file
> * time for a read after a backwards seek
> * time for a read after a forwards seek.
> * how many reads actually took place
> * for vector IO, whatever gets picked up there
> * were errors reported and retried, or throttling events
> * number of underlying GET requests
> 

CMIW, it seems that these stats can be collected solely at the input stream level.

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ColumnChunkPageReadStore.java
Comment (wgtmac) [reply parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ColumnChunkPageReadStore.java:83]: Did you forget to push your local changes? @parthchandra

Comment (parthchandra) [reply parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ColumnChunkPageReadStore.java]: Yes, I just pushed them now. 🤦🏾

**Comment by parthchandra** [comment]
> > For the object stores, things to measure are
> > 
> > * time to open() and close() a file
> > * time for a read after a backwards seek
> > * time for a read after a forwards seek.
> > * how many reads actually took place
> > * for vector IO, whatever gets picked up there
> > * were errors reported and retried, or throttling events
> > * number of underlying GET requests
> 
> CMIW, it seems that these stats can be collected solely at the input stream level.

Yes, they are best collected by the file system client API. However it would be nice to be able to hook up all these metrics together. Then we could, for instance, show a single Spark scan operator that displays stats for the operator, parquet reader, and the input stream in one place.

**Comment by wgtmac** [comment]
Thanks @parthchandra! Do you have any TODO work item on this (or after vectored I/O is merged)?

**Comment by parthchandra** [comment]
No I don't have any todo for this. If we are going to merge vector IO soon, I can add the metrics for that as well as part of this PR.

**Comment by wgtmac** [comment]
> No I don't have any todo for this. If we are going to merge vector IO soon, I can add the metrics for that as well as part of this PR.

No necessary. We can add them progressively.

@steveloughran @shangxinli @gszadovszky Do you have any concern to merge this?

**Comment by parthchandra** [comment]
@wgtmac do you think we can merge this?

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileReader.java
Comment (ConeyLiu) [inline parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileReader.java:135]: Why not set the `callback` in `ParquetReadOptions`?

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ColumnChunkPageReadStore.java
Comment (ConeyLiu) [reply parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ColumnChunkPageReadStore.java:80]: It would be better to mark it as final. I suggest getting `callback` from `ParquetReadOptions` and then we don't need to update the constructor and more reasonable.

**Comment by shangxinli** [comment]
LGTM

Thanks @parthchandra for working on this PR. And thanks @wgtmac and @ConeyLiu for reviewing the changes.

**Review by shangxinli — APPROVED** [review]

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileReader.java
Comment (parthchandra) [reply parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileReader.java:135]: That's a good idea, but it may require changes to some of the public constructors which are in use by others. For instance Spark has this https://github.com/apache/spark/blob/15e99cf676d9de02c54ca5ebe9a2bc6a3ce014e5/sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/SpecificParquetRecordReaderBase.java#L104
In this case, the constructor is deprecated and it looks like Spark could benefit from a constructor that takes a ParquetReadOptions as a parameter. I can introduce a new constructor but wasn't sure that we would want to do so. Having an additional API to set the metrics callback seemed like the simpler option.
WDYT?

Comment (wgtmac) [reply parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileReader.java:135]: @ConeyLiu Gentle ping, do you have any comment?

Comment (ConeyLiu) [reply parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileReader.java:135]: Sorry for missing this. https://github.com/apache/parquet-mr/blob/afd39dde8fd762bf696fea3dab16d45eae1093c3/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileReader.java#L669
The `open` method passing a `footer` is deprecated. I think we should support it with `ParquetReadOptions`. Then Spark could benefit from this as well. 

I am OK with the current changes to unblock this PR.

**Comment by ConeyLiu** [comment]
@parthchandra one more question, why reverted changes for https://github.com/apache/parquet-mr/pull/1187#discussion_r1396756877?

**Comment by parthchandra** [comment]
> @parthchandra one more question, why reverted changes for [#1187 (comment)](https://github.com/apache/parquet-mr/pull/1187#discussion_r1396756877)?

Hmm. I'm sure I didn't revert that. Let me address this.

Also, I think i will change this to use `ParquetReadOptions`, and add an additional `open` API for use by Spark. So let's hold off on merging this.

**Review by steveloughran — COMMENTED** [review]
just realised i had a comment but didn't submit it...here you go. Sorry

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetMetricsCallback.java
Comment (parthchandra) [reply parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetMetricsCallback.java:36]: > Can you tag as unstable but promise that all new callbacks will have a default implementation for any new metric

Done. 

Re: nanotimer - Learnt something new today. I remember when nanosecond timers were introduced things were unstable, but hopefully things are more reliable nowadays. I usually look at Guava's Stopwatch as the baseline implementation and the nanotimer implementation doesn't have a check for getting a negative diff either.  Anyhow, I added a check for that so we get at least zero.

**Review by parthchandra — COMMENTED** [review]
@ConeyLiu, @wgtmac, changed the implementation to use ParqutReadOptions and added a new constructor that takes a footer and read options as parameters (for use by Spark).
I checked Iceberg as well and it looks like it doesn't use any deprecated APIs.

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ColumnChunkPageReadStore.java
Comment (ConeyLiu) [inline parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ColumnChunkPageReadStore.java:93]: Why not get the `callback` by the `options.getMetricsCallback`?

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileReader.java
Comment (ConeyLiu) [inline parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileReader.java:831]: Will this eventually be replaced by another constructor or just not recommended?

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ColumnChunkPageReadStore.java
Comment (parthchandra) [reply parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ColumnChunkPageReadStore.java]: Of course. My oversight.

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileReader.java
Comment (parthchandra) [reply parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileReader.java]: Oops. Copied the old implementation and did not update the javadoc. Fixed.

**Comment by parthchandra** [comment]
Hi @ConeyLiu, @wgtmac, any additional concerns? If none, should we merge this?

**Review by ConeyLiu — APPROVED** [review]
+1, thanks @parthchandra  LGTM

**Comment by wgtmac** [comment]
Sorry for the delay. I just merged this. Thanks @parthchandra!

**Comment by parthchandra** [comment]
Thank you!

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetMetricsCallback.java
Comment (steveloughran) [reply parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetMetricsCallback.java:36]: +1
I should revisit state of the art there, and make sure that hadoop duration timers are good at -ve clock changes. [past commentary on 2015 CPUs](jhttps://steveloughran.blogspot.com/2015/09/time-on-multi-core-multi-socket-servers.html)

maybe something like some soft referenced "source of time" per thread which is both (systime, nanotime) and when querying it if current nanotime is "close enough" to previous nanotime, just apply nanotime diff to systime; update systime fields. So it'd be "approximate per cpu core/socket", enough for metrics where things happen in millis

**Pull request merged**
